src/widgets/widget_list.rs

```diff
@@ -0,0 +1,258 @@
+use crate::{
+    buffer::Buffer,
+    layout::{Corner, Rect, Size},
+    style::Style,
+    widgets::{Block, StatefulWidget, Widget},
+};
+
+use super::SizeHint;
+
+#[derive(Debug, Clone, Default)]
+pub struct WidgetListState {
+    offset: usize,
+    selected: Option<usize>,
+}
+
+impl WidgetListState {
+    pub fn selected(&self) -> Option<usize> {
+        self.selected
+    }
+
+    pub fn select(&mut self, index: Option<usize>) {
+        self.selected = index;
+        if index.is_none() {
+            self.offset = 0;
+        }
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct WidgetListItem<E: Widget + SizeHint + Clone> {
+    content: E,
+    style: Style,
+}
+
+impl<E> WidgetListItem<E>
+where
+    E: Widget + SizeHint + Clone,
+{
+    pub fn new(content: E) -> WidgetListItem<E> {
+        WidgetListItem {
+            content,
+            style: Style::default(),
+        }
+    }
+
+    pub fn style(mut self, style: Style) -> WidgetListItem<E> {
+        self.style = style;
+        self
+    }
```

Comment:
@joshka it is also not obvious for me  :smile: I really copy/pasted the list view and spent 2 hours making it work for the paragraph. I didn't expect also a full review, otherwise I would have spent some more time before posting the PR. but I will check this in the following days.

Reply:
Don't stress on this being a full review :) Seeing a WIP is great.
I'm jumping in on this purely because of personal interest on the implementation.

Reply:
for the moment I remove the style property, because it works only with text and paragraph widgets.
it is not really needed the user is still able to create different style to every item or the selected item. (in the widget_list example a different style is set to the selected widget )

Reply:
I put it back. because if padding is set, then you can set the background of the full list. it acts also as default  style for the items

Reply:
I think it's a good idea to leave it there - all the other widgets follow the pattern:

* set style to area
* draw block
* draw widget

And having that be predictable is probably a good thing. 

Aside - that pattern should be documented better, and changes to that are probably worth a larger separated discussion.